components/timezone/src/ids.rs

```diff
@@ -0,0 +1,371 @@
+    /// );
+    /// ```
+    pub fn bcp47_to_iana_search(&self, bcp47_id: TimeZoneBcp47Id) -> Option<String> {
+        // TODO: This is not as efficient as .probe() since it allocates a string each time.
+        for (string, raw_value) in self.data.map.iter() {
+            let trie_value = IanaTrieValue(raw_value);
+            if !trie_value.is_canonical() {
+                continue;
+            }
+            let Some(candidate_bcp47_id) = self.data.bcp47_ids.get(trie_value.index()) else {
+                debug_assert!(false, "index should be in range");
+                return None;
+            };
+            if candidate_bcp47_id == bcp47_id {
+                return Some(string);
+            }
+        }
+        None
+    }
+
+    fn iana_lookup_quick(&self, iana_id: &str) -> Option<IanaTrieValue> {
+        self.data.map.get(iana_id).map(IanaTrieValue)
+    }
+
+    fn iana_lookup_with_normalization<'l, 's>(
+        &'l self,
+        iana_id: &'s str,
+        mut cursor_fn: impl FnMut(&ZeroAsciiIgnoreCaseTrieCursor<'l>),
+    ) -> Option<(IanaTrieValue, Cow<'s, str>)> {
+        let mut cursor = self.data.map.cursor();
+        let mut string = Cow::Borrowed(iana_id);
+        let mut i = 0;
+        let trie_value = loop {
+            cursor_fn(&cursor);
+            let Some(input_byte) = string.as_bytes().get(i).copied() else {
+                break cursor.take_value().map(IanaTrieValue);
+            };
+            let Some(matched_byte) = cursor.step(input_byte) else {
+                break None;
+            };
+            if matched_byte != input_byte {
+                // Safety: we write to input_byte farther down after performing safety checks.
+                let Some(input_byte) = unsafe { string.to_mut().as_bytes_mut() }.get_mut(i) else {
+                    debug_assert!(false, "the same index was just accessed earlier");
+                    break None;
+                };
+                if !input_byte.is_ascii() {
+                    debug_assert!(false, "non-ASCII input byte: {input_byte}");
+                    break None;
+                }
+                if !matched_byte.is_ascii() {
+                    debug_assert!(false, "non-ASCII matched byte: {matched_byte}");
+                    break None;
+                }
+                // Safety: we just checked that both input_byte and matched_byte are ASCII,
+                // so the buffer remains UTF-8 when we replace one with the other.
+                *input_byte = matched_byte;
+            }
+            i += 1;
+        }?;
+        Some((trie_value, string))
+    }
+}
+
+/// A wrapper around a syntax-normalized IANA time zone identifier string
+/// and its corresponding BCP-47 time zone identifier.
+#[derive(Debug)]
+pub struct NormalizedIana<'s> {
+    /// The syntax-normalized IANA time zone identifier string.
+    pub string: Cow<'s, str>,
+    /// The corresponding BCP-47 time zone identifier.
+    pub bcp47_id: TimeZoneBcp47Id,
```

Comment:
issue: I don't think this is the correct place for it. Not every string that's normalized according to the IANA syntax has a bcp47 id. Also, having a `pub string` allows the creation of unnormalized `NormalizedIana`s.

I'd suggest

```rust
pub struct NormalizedIanaId<'s>(Cow<'s, str>);
// derefs to str
```

And then the canonicalize operation can return `(NormalizedIanaId, TimeZoneBcp47Id)`.

Reply:
Actually, maybe the names `(TimeZoneIanaId, TimeZoneBcp47Id)` are more symmetric. According to our style guide it should probably have a `TimeZone` prefix. The normalization doesn't necessarily need to be called out explicitly, same way it's not for `Locale` et al.

Reply:
This is a good observation. The difference is that currently `TimeZoneBcp47Id` is an open enum (newtype with no invariants) whereas `NormalizedIana` enforces validity. I'm open to discussing this.

Reply:
I also stuffed the BCP-47 ID into `NormalizedIana` as another field, so `TimeZoneIanaId` isn't really the right name for this specific spot; we might want to use it later for something that is actually the analog of `TimeZoneBcp47Id`.

Reply:
My original comment also proposed moving out the `bcp47_id` field, and returning tuples if you need to return both.

Reply:
Ok, I removed `NormalizedIana` and replaced them with tuples of the correct types. Sometimes I return `Cow<str>`, sometimes `String`, sometimes `&str`. Sometimes I return the BCP-47, if it is there, and sometimes not.